### PR TITLE
P2 681 add inherit sidebar block

### PR DESCRIPTION
**Pull request by increddibelly**
## Summary

This PR can be summarized in the following changelog entry:

* [@yoast/schema-blocks] Adds Inherit-Sidebar instruction.

## Relevant technical choices:

* added an inherit-sidebar instruction which allows subblocks to show the selected parents' sidebars. 
## Test instructions

This PR can be tested by following these steps:

* add `{{ inherit-sidebar }}` to the template that resides in an innerblock of another template; e.g. add this to the `steps` block template of the `yoast/recipe` block template (the premium pr for p2-681 already does this for the job blocks)
* add the template in a new post; select any of the job posting subblocks from the template. Notice the sidebar automatically opens.
* navigate between the innerblock and parent block; notice how the block name (top right) changes to indicate the current block, but the sidebar for the parent block remains.
* add a core/paragraph block to any of the job posting subblocks; notice how the sidebar changes to the core/paragraph sidebar.

## Impact check

* This PR affects the following parts of the plugin, which may require extra testing:
  *

## UI changes
* [x] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes #

### CODE REVIEW

File: packages/schema-blocks/src/instructions/blocks/InheritSidebar.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+			if ( parentId ) {
+				parentIds = [ parentId ];
+			}
+		}
+
+		const elements: ReactElement[] = [];
+		if ( parentIds.length > 0 ) {
+			parentIds.forEach( parentId => {
+				const parentBlock = getBlockByClientId( parentId );
+				const parentBlockInstruction = Instruction.getBlock( parentBlock.name );
```
Comment (increddibelly) [inline packages/schema-blocks/src/instructions/blocks/InheritSidebar.tsx:45]: trying to get the registered instruction with its options from the registry; this does not work as expected.

File: packages/schema-blocks/src/core/Instruction.ts [hunk 2]
```diff
@@ -79,6 +80,16 @@ export default abstract class Instruction {
 		this.registeredInstructions[ name ] = instruction;
 	}
 
+	/**
+	 * Retrieves a registered blockinstruction.
+	 * @param this This.
+	 * @param name The block to retrieve.
+	 * @returns {InstructionClass} The registered BlockInstruction.
+	 */
+	static getBlock<I extends typeof Instruction>( this: I, name: string ): InstructionClass<I["prototype"]> {
```
Comment (increddibelly) [inline packages/schema-blocks/src/core/Instruction.ts:89]: unused method - removed.

File: packages/schema-blocks/src/core/blocks/BlockDefinition.tsx [hunk 3]
```diff
@@ -89,4 +92,21 @@ export default class BlockDefinition extends Definition {
...
+	 * Creates the sidebar elements.
+	 *
+	 * @param props The properties of the block to create a sidebar for.
+	 *
+	 * @returns {ReactElement[]} The sidebar element to render.
+	 */
+	sidebarElements( props: RenderEditProps ): ReactElement[] {
+		const sidebarElements: ReactElement[] = Object.values( this.instructions )
+			.map( ( instruction, index ) => {
+				const output = instruction.sidebar( props, index );
```
Comment (increddibelly) [inline packages/schema-blocks/src/core/blocks/BlockDefinition.tsx:106]: I removed the intermediate variables

File: packages/schema-blocks/src/functions/gutenberg/sidebar.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { select, dispatch } from "@wordpress/data";
+
+/**
+ * Opens the given sidebar unless the given sidebar or another sidebar is already open.
```
Comment (andizer) [inline packages/schema-blocks/src/functions/gutenberg/sidebar.ts:4]: Not catched by the linter, but I would love to see a newline after this.

```suggestion
 * Opens the given sidebar unless the given sidebar or another sidebar is already open.
 *
```

File: packages/schema-blocks/src/core/blocks/BlockInstruction.ts [hunk 5]
```diff
@@ -51,17 +54,17 @@ export default abstract class BlockInstruction extends Instruction {
 	 * @param props The props.
 	 * @param i     The number the rendered element is of its parent.
 	 *
-	 * @returns The sidebar element to render.
+	 * @returns {ReactElement | string} The sidebar element to render.
```
Comment (andizer) [inline packages/schema-blocks/src/core/blocks/BlockInstruction.ts:57]: It now returns a ReactElement right? I think you can remove the `| string` part?
```suggestion
	 * @returns {ReactElement} The sidebar element to render.
```

File: packages/schema-blocks/src/core/blocks/BlockDefinition.tsx [hunk 6]
```diff
@@ -89,4 +92,21 @@ export default class BlockDefinition extends Definition {
...
+				return output;
```
Comment (andizer) [inline packages/schema-blocks/src/core/blocks/BlockDefinition.tsx:107]: `output` is returned immediately after its declaration, I'd just return the result of the function call without storing it to a variable
```suggestion
				return instruction.sidebar( props, index );
```

File: packages/schema-blocks/src/instructions/blocks/InheritSidebar.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+	/* eslint-disable @typescript-eslint/no-unused-vars */
+	/**
+	 * Renders the value of a sidebar input.
+	 *
+	 * @param props The render props.
+	 *
+	 * @returns The value of the sidebar input.
+	 */
+	protected value( props: RenderSaveProps | RenderEditProps ): string {
```
Comment (andizer) [inline packages/schema-blocks/src/instructions/blocks/InheritSidebar.tsx:79]: This methods isn't used? Also the `BlockInstruction` doesn't seem to have a `value` method. I would remove it.

File: packages/schema-blocks/src/functions/gutenberg/block.ts [hunk 8]
```diff
@@ -33,3 +33,49 @@ export function getBlockId( block: BlockInstance ): string {
...
+ */
+type extendedCoreBlockEditorSelector = {
+	getBlockParentsByBlockName( clientId: string, parentNames: string[] ): string[];
+}
+
+/**
+ * Creates BlockEditProps for a block with a given blockId.
+ * @param block The block to create BlockEditProps for.
+ * @param selected True if the block is selected in the editor; default false.
+ * @returns {BEP} The BlockEditProps.
```
Comment (increddibelly) [inline packages/schema-blocks/src/functions/gutenberg/block.ts:71]: BlockEditProps*

**Review by andizer — CHANGES_REQUESTED** [review]
CR done

File: packages/schema-blocks/src/core/blocks/BlockDefinition.tsx (hunk 6 above)
Comment (increddibelly) [reply packages/schema-blocks/src/core/blocks/BlockDefinition.tsx]: the entire function is now a oneliner.

File: packages/schema-blocks/src/core/blocks/BlockInstruction.ts (hunk 5 above)
Comment (increddibelly) [reply packages/schema-blocks/src/core/blocks/BlockInstruction.ts:57]: this breaks inherited blocks that DO return a string. so, not for now, unfortunately.

File: packages/schema-blocks/src/functions/gutenberg/sidebar.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+ * @param sidebarName The sidebar to open, default "yoast-seo/seo-sidebar".
+ * @param force If true, closes the currently opened sidebar and opens the given sidebar, unless the current sidebar is the requested sidebar.
+ */
+export function openGeneralSidebar( sidebarName = "yoast-seo/seo-sidebar", force = false ): void {
+	const isEditorSidebarOpened = select( "core/edit-post" ).isEditorSidebarOpened();
+	const activeGeneralSidebarName = select( "core/edit-post" ).getActiveGeneralSidebarName();
+
+	if ( ( force || ! isEditorSidebarOpened ) && ( activeGeneralSidebarName !== sidebarName ) ) {
```
[the remark this answers is not in the document]
Comment (increddibelly) [reply packages/schema-blocks/src/functions/gutenberg/sidebar.ts:12]: indeed. removed the double negation.

File: packages/schema-blocks/src/instructions/blocks/Block.tsx [hunk 10]
```diff
@@ -37,38 +34,9 @@ class Block extends BlockInstruction {
 	 * @returns The sidebar element to render.
 	 */
 	sidebar( props: RenderEditProps ): ReactElement {
```
[the remark this answers is not in the document]
Comment (increddibelly) [reply packages/schema-blocks/src/instructions/blocks/Block.tsx:36]: ooh, will have a look!

**Comment by igorschoester** [comment]
Please rebase this to where it needs to go. The `release/15.9` branch is no longer valid, it has been released already.

**Review by andizer — APPROVED** [review]
CR 👍

**Comment by hansjovis** [comment]
Acceptance test: ✅

**Pull request merged**